### ISSUE TITLE
Highlight selection in a single long line is very slow

### ISSUE THREAD

**Issue by EFanZh** [Microsoft/vscode#13515]
- VSCode Version: 1.6.0
- OS Version: Windows 10

Steps to Reproduce:

1. Create an empty file;
2. Copy this line into your editor:

        aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
3. Select a single character ‘a’;
4. Observe the response time.

If it is not slow enough for you, you can try with more characters.

The response time is tolerable (but still a little long) if characters are distributed in different lines. Try this:

```
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```

This one is a lot faster than the previous one although they have the same number of ‘a’s.

Maybe the problem is with the algorithm. If the response time can’t be reduced, can you do the highlighting work in the background? I don’t want any thing to block my editing process. I don’t mind if some highlighting is delayed, but I can’t stand the UI being unresponsive.

**Comment by alexandrudima** [comment]
@EFanZh 
### A bit of context on what we do:
* we render the line with the text (this is pretty fast)
* when selecting one `a`
* there is a `SelectionHighlighter` that can be turned off via `"editor.selectionHighlight": false`
* it adds a decoration to each other `a` in the file (with a limit of 1000)
* to render decorations, we don't want to make assumptions about text layout
* (the browser does the text layout, i.e. bidi text, non-monospace font, etc.)
* for each decoration we want to render we ask the browser where the range ended up
* i.e. for the first `a`, we ask the browser where did end up painted, for the second, ... and so on
* since this is a single token, it ends up in a single `<span>` node
* I believe Chromium has a `O(N)` algorithm for replying to our queries
* so basically you get `O(N^2)` where N is capped at 1000
* even on my quite good desktop machine this results in 185ms time spent to render a single frame

### Already an improvement on the insiders release channel
* a couple of days ago, I've added some optimizations:
* if the text is guaranteed to not contain RTL or bidi (i.e. does not contain Unicode characters that are in the "R" or "AL" category)
* then we build up [a cache of locations for characters](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/browser/viewParts/lines/viewLine.ts#L253)
* and we can assume that the decoration forms a contiguous region (i.e. from left to right)
* this already brings down the first frame rendering from 185ms to 69ms
* subsequent frames (until you touch the line's text) (thanks to the cache) render in less than 10ms

### Possible future improvement A)
* we could artificially break the text into multiple `<span>`s
* I think browsers cache the layout of each `<span>` (i.e. where the span sits), but they don't cache the layout of the TextNode, therefore the long time spent in answering our queries.
* this is somewhat confirmed by distributing the same text on multiple lines

### Possible future improvement B)
* similar to how we have an heuristic that tells us the text is guaranteed to be LTR
* we could have a heuristic to detect if the font is monospace
* if the font is monospace and the text is LTR, we can skip the reading entirely and replace it with a simple mutliplication (taking into account CJK characters as well, or perhaps only bring the optimization in for ASCII)

### Long story short
@EFanZh you can try out the insiders channel or temporarily disable `"editor.selectionHighlight": false` until we get decorations rendering a lot faster.

**Comment by alexandrudima** [comment]
Related #6885

**Comment by alexandrudima** [comment]
Improvement A) is implemented via c0def7092d0e5254241d2f2ac97184ad1f024487

**Comment by alexandrudima** [comment]
Improvement B) is now implemented via the `FastRenderedViewLine`.
